independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java

```diff
@@ -0,0 +1,85 @@
+package io.quarkus.qute;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Escapes a characted sequence using a map of replacements.
+ */
+public final class Escaper {
+
+    private final Map<Character, String> replacements;
+
+    /**
+     *
+     * @param replacements
+     */
+    private Escaper(Map<Character, String> replacements) {
+        this.replacements = replacements.isEmpty() ? Collections.emptyMap()
+                : new HashMap<>(replacements);
+    }
+
+    /**
+     *
+     * @param value
+     * @return an escaped value
+     */
+    public String escape(CharSequence value) {
+        Objects.requireNonNull(value);
+        if (value.length() == 0) {
+            return value.toString();
+        }
+        for (int i = 0; i < value.length(); i++) {
+            String replacement = replacements.get(value.charAt(i));
+            if (replacement != null) {
+                // In most cases we will not need to escape the value at all
+                return doEscape(value, i, new StringBuilder(value.subSequence(0, i)).append(replacement));
+            }
+        }
+        return value.toString();
+    }
+
+    private String doEscape(CharSequence value, int index, StringBuilder builder) {
+        int length = value.length();
+        while (++index < length) {
+            char c = value.charAt(index);
```

Comment:
Watch out that this is not codepoint-compliant. Not sure if 16-bit components of 32-bit codepoints can be in the ascii plane, but we should check.

Reply:
I have no idea what you're talking about ;-). Chars in Java are unicode, or? I would be more than grateful if you could check whatever needs to be checked.

Reply:
chars in java are "half" unicode. they're 16 bits, so for any "unicode char" (which is called a "codepoint") that requires more than 16 bits they're encoded on two consecutive 16-bit chars using UTF-16 (standard encoding of all Java Strings at runtime).

Reply:
Thus iterating chars for comparison is rarely right: you must use the codepoint static methods on `Character` to ask if a codepoint is on 1 or 2 chars. Now, it's possible that the 2-char codepoints have both chars greater than the ascii range, and you're only escaping (5) ascii values so your code may work, but we should check.

Reply:
Both surrogate pairs have their sign bits at 1 so they can't be mistaken for the ascii we're replacing, so we're good.